<details>
<summary><code>container_inspect_linux_test.go:447: assertion failed: on-failure (string) !=  (inspect.HostConfig.RestartPolicy.Name string)</code></summary>
<p>

```console
$ go test -exec sudo -run '^TestContainerInspectHostConfig$' ./cmd/nerdctl/container/ 
test target: "nerdctl"
--- FAIL: TestContainerInspectHostConfig (1.25s)
    container_inspect_linux_test.go:458: 
        
        +============================================================================================================+
        | 🧽      | "TestContainerInspectHostConfig": initial cleanup                                                |
        +============================================================================================================+
        
    container_inspect_linux_test.go:411: 
        
        +------------------------------------------------------------------------------------------------------------+
        | ➡️      | ⚙️ /usr/local/bin/nerdctl rm -f testcontainerinspecthostconfig-92ba1d6f                          |
        +------------------------------------------------------------------------------------------------------------+
        |         | 🟠 time="2026-06-19T13:15:35+09:00" level=error msg="1 errors:\nno such container: testcontainer |
        |         | inspecthostconfig-92ba1d6f"                                                                      |
        +------------------------------------------------------------------------------------------------------------+
        | 🌱      | HOME=/root                                                                                       |
        |         | PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin    |
        |         | DOCKER_CONFIG=/tmp/TestContainerInspectHostConfig3131155204/003                                  |
        |         | NERDCTL_TOML=/tmp/TestContainerInspectHostConfig3131155204/003/nerdctl.toml                      |
        +------------------------------------------------------------------------------------------------------------+
        | ⏰      | <1s (limit: 3m0s)                                                                                |
        +------------------------------------------------------------------------------------------------------------+
        | 📁      | /tmp/TestContainerInspectHostConfig3131155204/002                                                |
        +------------------------------------------------------------------------------------------------------------+
        
    container_inspect_linux_test.go:458: 
        
        +============================================================================================================+
        | 🏗       | "TestContainerInspectHostConfig": setup                                                          |
        +============================================================================================================+
        
    container_inspect_linux_test.go:385: 
        
        +------------------------------------------------------------------------------------------------------------+
        | ➡️      | ⚙️ /usr/local/bin/nerdctl run -d --name testcontainerinspecthostconfig-92ba1d6f --cpuset-cpus 0- |
        |         | 1 --cpuset-mems 0 --cpu-shares 1024 --cpu-quota 100000 --group-add 1000 --group-add 2000 --add-h |
        |         | ost host1:10.0.0.1 --add-host host2:10.0.0.2 --ipc host --memory 512m --memory-reservation 200m  |
        |         | --memory-swappiness 60 --pids-limit 100 --ulimit nofile=1024:65536 --read-only --shm-size 256m - |
        |         | -uts host --restart on-failure:3 --runtime io.containerd.runc.v2 --annotation com.example.key=te |
        |         | st-val ghcr.io/stargz-containers/alpine:3.13-org sleep 3600                                      |
        +------------------------------------------------------------------------------------------------------------+
        |         | 🟢 c9cbd0900dcb1ed237baf1243bb809e6438fb118dbb747261469d4864378ced8                              |
        +------------------------------------------------------------------------------------------------------------+
        | 🌱      | HOME=/root                                                                                       |
        |         | PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin    |
        |         | DOCKER_CONFIG=/tmp/TestContainerInspectHostConfig3131155204/003                                  |
        |         | NERDCTL_TOML=/tmp/TestContainerInspectHostConfig3131155204/003/nerdctl.toml                      |
        +------------------------------------------------------------------------------------------------------------+
        | ⏰      | <1s (limit: 3m0s)                                                                                |
        +------------------------------------------------------------------------------------------------------------+
        | 📁      | /tmp/TestContainerInspectHostConfig3131155204/002                                                |
        +------------------------------------------------------------------------------------------------------------+
        
    container_inspect_linux_test.go:407: 
        
        +------------------------------------------------------------------------------------------------------------+
        | ➡️      | ⚙️ /usr/local/bin/nerdctl container inspect testcontainerinspecthostconfig-92ba1d6f              |
        +------------------------------------------------------------------------------------------------------------+
        |         | Actual content is 195 lines long and has been abbreviated to 50
         |
        |         |                                                                                                  |
        |         | 🟢 [                                                                                             |
        |         |     {                                                                                            |
        |         |         "Id": "c9cbd0900dcb1ed237baf1243bb809e6438fb118dbb747261469d4864378ced8",                |
        |         |         "Created": "2026-06-19T04:15:35.600363949Z",                                             |
        |         |         "Path": "sleep",                                                                         |
        |         |         "Args": [                                                                                |
        |         |             "3600"                                                                               |
        |         |         ],                                                                                       |
        |         |         "State": {                                                                               |
        |         |             "Status": "running",                                                                 |
        |         |             "Running": true,                                                                     |
        |         |             "Paused": false,                                                                     |
        |         | ... 145 lines are being ignored...                                                               |
        |         |                 "nerdctl/dns": "{\"DNSServers\":null,\"DNSResolvConfOptions\":null,\"DNSSearchDo |
        |         | mains\":null}",                                                                                  |
        |         |                 "nerdctl/extraHosts": "[\"host1:10.0.0.1\",\"host2:10.0.0.2\"]",                 |
        |         |                 "nerdctl/host-config": "{\"BlkioWeight\":0,\"CidFile\":\"\",\"Devices\":null}",  |
        |         |                 "nerdctl/ipc": "{\"mode\":\"host\",\"shmSize\":\"256m\"}",                       |
        |         |                 "nerdctl/log-config": "{\"driver\":\"json-file\",\"address\":\"/run/containerd/c |
        |         | ontainerd.sock\"}",                                                                              |
        |         |                 "nerdctl/log-uri": "binary:///usr/local/bin/nerdctl?_NERDCTL_INTERNAL_LOGGING=%2 |
        |         | Fvar%2Flib%2Fnerdctl%2F1935db59",                                                                |
        |         |                 "nerdctl/name": "testcontainerinspecthostconfig-92ba1d6f",                       |
        |         |                 "nerdctl/namespace": "nerdctl-test",                                             |
        |         |                 "nerdctl/networks": "[\"bridge\"]",                                              |
        |         |                 "nerdctl/platform": "linux/arm64/v8",                                            |
        |         |                 "nerdctl/state-dir": "/var/lib/nerdctl/1935db59/containers/nerdctl-test/c9cbd090 |
        |         | 0dcb1ed237baf1243bb809e6438fb118dbb747261469d4864378ced8"                                        |
        |         |             }                                                                                    |
        |         |         },                                                                                       |
        |         |         "NetworkSettings": {                                                                     |
        |         |             "Ports": {},                                                                         |
        |         |             "GlobalIPv6Address": "",                                                             |
        |         |             "GlobalIPv6PrefixLen": 0,                                                            |
        |         |             "IPAddress": "10.4.0.23",                                                            |
        |         |             "IPPrefixLen": 24,                                                                   |
        |         |             "MacAddress": "ea:40:a7:2a:24:af",                                                   |
        |         |             "Networks": {                                                                        |
        |         |                 "unknown-eth0": {                                                                |
        |         |                     "IPAddress": "10.4.0.23",                                                    |
        |         |                     "IPPrefixLen": 24,                                                           |
        |         |                     "GlobalIPv6Address": "",                                                     |
        |         |                     "GlobalIPv6PrefixLen": 0,                                                    |
        |         |                     "MacAddress": "ea:40:a7:2a:24:af"                                            |
        |         |                 }                                                                                |
        |         |             }                                                                                    |
        |         |         }                                                                                        |
        |         |     }                                                                                            |
        |         | ]                                                                                                |
        |         |                                                                                                  |
        +------------------------------------------------------------------------------------------------------------+
        | 🌱      | HOME=/root                                                                                       |
        |         | PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin    |
        |         | DOCKER_CONFIG=/tmp/TestContainerInspectHostConfig3131155204/003                                  |
        |         | NERDCTL_TOML=/tmp/TestContainerInspectHostConfig3131155204/003/nerdctl.toml                      |
        +------------------------------------------------------------------------------------------------------------+
        | ⏰      | <1s (limit: 3m0s)                                                                                |
        +------------------------------------------------------------------------------------------------------------+
        | 📁      | /tmp/TestContainerInspectHostConfig3131155204/002                                                |
        +------------------------------------------------------------------------------------------------------------+
        
    container_inspect_linux_test.go:458: 
        
        +============================================================================================================+
        | 🚀      | "TestContainerInspectHostConfig": starting test!                                                 |
        +============================================================================================================+
        | ⏳      | /tmp/TestContainerInspectHostConfig3131155204/001/c9f902e1                                       |
        +============================================================================================================+
        |         |                                                                                                  |
        +============================================================================================================+
        | config  | {                                                                                                |
        |         |   "Namespace": "nerdctl-test",                                                                   |
        |         |   "PrivateMode": ""                                                                              |
        |         | }                                                                                                |
        +============================================================================================================+
        | labels  | {}                                                                                               |
        +============================================================================================================+
        
    container_inspect_linux_test.go:458: 
        
        +------------------------------------------------------------------------------------------------------------+
        | ➡️      | ⚙️ /usr/local/bin/nerdctl inspect testcontainerinspecthostconfig-92ba1d6f                        |
        +------------------------------------------------------------------------------------------------------------+
        |         | Actual content is 195 lines long and has been abbreviated to 50
         |
        |         |                                                                                                  |
        |         | 🟢 [                                                                                             |
        |         |     {                                                                                            |
        |         |         "Id": "c9cbd0900dcb1ed237baf1243bb809e6438fb118dbb747261469d4864378ced8",                |
        |         |         "Created": "2026-06-19T04:15:35.600363949Z",                                             |
        |         |         "Path": "sleep",                                                                         |
        |         |         "Args": [                                                                                |
        |         |             "3600"                                                                               |
        |         |         ],                                                                                       |
        |         |         "State": {                                                                               |
        |         |             "Status": "running",                                                                 |
        |         |             "Running": true,                                                                     |
        |         |             "Paused": false,                                                                     |
        |         | ... 145 lines are being ignored...                                                               |
        |         |                 "nerdctl/dns": "{\"DNSServers\":null,\"DNSResolvConfOptions\":null,\"DNSSearchDo |
        |         | mains\":null}",                                                                                  |
        |         |                 "nerdctl/extraHosts": "[\"host1:10.0.0.1\",\"host2:10.0.0.2\"]",                 |
        |         |                 "nerdctl/host-config": "{\"BlkioWeight\":0,\"CidFile\":\"\",\"Devices\":null}",  |
        |         |                 "nerdctl/ipc": "{\"mode\":\"host\",\"shmSize\":\"256m\"}",                       |
        |         |                 "nerdctl/log-config": "{\"driver\":\"json-file\",\"address\":\"/run/containerd/c |
        |         | ontainerd.sock\"}",                                                                              |
        |         |                 "nerdctl/log-uri": "binary:///usr/local/bin/nerdctl?_NERDCTL_INTERNAL_LOGGING=%2 |
        |         | Fvar%2Flib%2Fnerdctl%2F1935db59",                                                                |
        |         |                 "nerdctl/name": "testcontainerinspecthostconfig-92ba1d6f",                       |
        |         |                 "nerdctl/namespace": "nerdctl-test",                                             |
        |         |                 "nerdctl/networks": "[\"bridge\"]",                                              |
        |         |                 "nerdctl/platform": "linux/arm64/v8",                                            |
        |         |                 "nerdctl/state-dir": "/var/lib/nerdctl/1935db59/containers/nerdctl-test/c9cbd090 |
        |         | 0dcb1ed237baf1243bb809e6438fb118dbb747261469d4864378ced8"                                        |
        |         |             }                                                                                    |
        |         |         },                                                                                       |
        |         |         "NetworkSettings": {                                                                     |
        |         |             "Ports": {},                                                                         |
        |         |             "GlobalIPv6Address": "",                                                             |
        |         |             "GlobalIPv6PrefixLen": 0,                                                            |
        |         |             "IPAddress": "10.4.0.23",                                                            |
        |         |             "IPPrefixLen": 24,                                                                   |
        |         |             "MacAddress": "ea:40:a7:2a:24:af",                                                   |
        |         |             "Networks": {                                                                        |
        |         |                 "unknown-eth0": {                                                                |
        |         |                     "IPAddress": "10.4.0.23",                                                    |
        |         |                     "IPPrefixLen": 24,                                                           |
        |         |                     "GlobalIPv6Address": "",                                                     |
        |         |                     "GlobalIPv6PrefixLen": 0,                                                    |
        |         |                     "MacAddress": "ea:40:a7:2a:24:af"                                            |
        |         |                 }                                                                                |
        |         |             }                                                                                    |
        |         |         }                                                                                        |
        |         |     }                                                                                            |
        |         | ]                                                                                                |
        |         |                                                                                                  |
        +------------------------------------------------------------------------------------------------------------+
        | 🌱      | HOME=/root                                                                                       |
        |         | PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin    |
        |         | DOCKER_CONFIG=/tmp/TestContainerInspectHostConfig3131155204/003                                  |
        |         | NERDCTL_TOML=/tmp/TestContainerInspectHostConfig3131155204/003/nerdctl.toml                      |
        +------------------------------------------------------------------------------------------------------------+
        | ⏰      | <1s (limit: 3m0s)                                                                                |
        +------------------------------------------------------------------------------------------------------------+
        | 📁      | /tmp/TestContainerInspectHostConfig3131155204/002                                                |
        +------------------------------------------------------------------------------------------------------------+
        
    container_inspect_linux_test.go:447: assertion failed: on-failure (string) !=  (inspect.HostConfig.RestartPolicy.Name string)
    container_inspect_linux_test.go:458: 
        
        +============================================================================================================+
        | 🧽      | "TestContainerInspectHostConfig": post-cleanup                                                   |
        +============================================================================================================+
        
    container_inspect_linux_test.go:411: 
        
        +------------------------------------------------------------------------------------------------------------+
        | ➡️      | ⚙️ /usr/local/bin/nerdctl rm -f testcontainerinspecthostconfig-92ba1d6f                          |
        +------------------------------------------------------------------------------------------------------------+
        |         | 🟢 testcontainerinspecthostconfig-92ba1d6f                                                       |
        +------------------------------------------------------------------------------------------------------------+
        | 🌱      | HOME=/root                                                                                       |
        |         | PATH=/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/snap/bin    |
        |         | DOCKER_CONFIG=/tmp/TestContainerInspectHostConfig3131155204/003                                  |
        |         | NERDCTL_TOML=/tmp/TestContainerInspectHostConfig3131155204/003/nerdctl.toml                      |
        +------------------------------------------------------------------------------------------------------------+
        | ⏰      | <1s (limit: 3m0s)                                                                                |
        +------------------------------------------------------------------------------------------------------------+
        | 📁      | /tmp/TestContainerInspectHostConfig3131155204/002                                                |
        +------------------------------------------------------------------------------------------------------------+
        
FAIL
FAIL	github.com/containerd/nerdctl/v2/cmd/nerdctl/container	1.256s
FAIL
```

</p>
</details>

Caused by [Use client.WithImageConfigLabels for image config labels](https://github.com/containerd/nerdctl/pull/4982/changes/7a2a1648b321506a6d92069e72e707a718bd5742) 7a2a1648b321506a6d92069e72e707a718bd5742, not by [go.mod: github.com/containerd/containerd/v2 v2.3.2](https://github.com/containerd/nerdctl/pull/4982/changes/87a857f8f5b3197efa45aaeacab4fcba07cb893a) 87a857f8f5b3197efa45aaeacab4fcba07cb893a

--- a/pkg/cmd/container/create.go
+++ b/pkg/cmd/container/create.go
@@ -206,6 +206,12 @@ func Create(ctx context.Context, client *containerd.Client, args []string, netMa
 	if err != nil {
 		return nil, generateRemoveStateDirFunc(ctx, id, internalLabels), err
 	}
+	// Image config labels must be applied before any other label-setting opt:
+	// containerd.WithImageConfigLabels resets the container labels, so running it
+	// later would clear labels set by other opts (e.g. the restart policy).
+	if ensuredImage != nil {
+		cOpts = append(cOpts, containerd.WithImageConfigLabels(ensuredImage.Image))
+	}
 	opts = append(opts, rootfsOpts...)
 	cOpts = append(cOpts, rootfsCOpts...)
 	if options.UserNS != "" {
@@ -357,7 +363,7 @@ func Create(ctx context.Context, client *containerd.Client, args []string, netMa
 		internalLabels.healthcheck = healthcheckConfig
 	}
 
-	lCOpts, err := withContainerLabels(options.Label, options.LabelFile, ensuredImage)
+	lCOpts, err := withContainerLabels(options.Label, options.LabelFile)
 	if err != nil {
 		return nil, generateRemoveOrphanedDirsFunc(ctx, id, dataStore, internalLabels), err
 	}
@@ -650,14 +656,8 @@ func withNerdctlOCIHook(cmd string, args []string) (oci.SpecOpts, error) {
 	}, nil
 }
 
-func withContainerLabels(label, labelFile []string, ensuredImage *imgutil.EnsuredImage) ([]containerd.NewContainerOpts, error) {
+func withContainerLabels(label, labelFile []string) ([]containerd.NewContainerOpts, error) {
 	var opts []containerd.NewContainerOpts
-
-	// add labels defined by image
-	if ensuredImage != nil {
-		imageLabelOpts := containerd.WithAdditionalContainerLabels(ensuredImage.ImageConfig.Labels)
-		opts = append(opts, imageLabelOpts)
-	}
 
 	labelMap, err := readKVStringsMapfFromLabel(label, labelFile)
 	if err != nil {

